symphony/bdk/core/service/datafeed/abstract_datafeed_loop.py

```diff
@@ -72,30 +73,30 @@ def __init__(self, datafeed_api: DatafeedApi, auth_session: AuthSession, config:
         self.bdk_config = config
         self.api_client = datafeed_api.api_client
         self.running = False
+        self.tasks = []
+        self.hard_kill = False
 
     async def start(self):
         """Start the datafeed event service
 
         :return: None
         """
+        logger.debug("Starting datafeed loop")
         await self.prepare_datafeed()
-        self.running = True
-        while self.running:
-            try:
-                event_list = await self.read_datafeed()
-                await self.handle_v4_event_list(event_list)
-            except ApiException as exc:
-                if exc.status == 400:
-                    await self.recreate_datafeed()
-                else:
-                    raise exc
-
-    async def stop(self):
+        await self._run_loop()
+
+        logger.debug("Stopping datafeed loop")
+        await self._stop_listener_tasks()
+
+    async def stop(self, hard_kill=False):
         """Stop the datafeed event service
 
+        :param hard_kill: if set to True, tasks running listener methods will be cancelled immediately. Otherwise, tasks
+          will be awaited until completion.
         :return: None
         """
         self.running = False
+        self.hard_kill = hard_kill
```

Comment:
is there a chance that the loop stops before we set the hard_kill value?

Reply:
Loop can stop if there is an exception inside the `while` loop, typically when calling `await self.read_datafeed()` if it throws a non 400 error. In this case, exception is propagated.
Indeed, we have to handle that case.

Reply:
Refactored the loop: params put in `start` method, finally block added